File: backend/registrations/checkIn.mjs
import { GetCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { db, REGISTRATIONS_TABLE } from "./shared/db.mjs";
import { ok, badRequest, notFound, serverError, cors } from "./shared/response.mjs";
import { audit, callerFromEvent } from "./shared/auth.mjs";

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const caller = callerFromEvent(event);
  const { id } = event.pathParameters ?? {};
  if (!id) return badRequest("id is required");

  try {
    const existing = await db.send(new GetCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
    }));
    if (!existing.Item) return notFound("Registration not found");
    if (existing.Item.checkedInAt) return ok({ ...existing.Item, alreadyCheckedIn: true });

    const now = new Date().toISOString();
    await db.send(new UpdateCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
      UpdateExpression: "SET checkedInAt = :t, checkedInBy = :by, updatedAt = :t",
      ExpressionAttributeValues: {
        ":t": now,
        ":by": caller.actorId ?? "system",
      },
    }));
    await audit("participant.checked_in", {
      entity: "participant",
      entityId: id,
      actorId: caller.actorId,
      actorLabel: caller.actorLabel,
      meta: { name: existing.Item.fullName, email: existing.Item.email, phone: existing.Item.phone },
    });
    return ok({ ...existing.Item, checkedInAt: now, checkedInBy: caller.actorId });
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/registrations/checkIn.py
import logging

from shared.auth import audit, caller_from_event, iso_now
from shared.db import registrations_table
from shared.response import bad_request, cors, not_found, ok, server_error, with_cors

log = logging.getLogger()


@with_cors
def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    caller = caller_from_event(event)
    reg_id = (event.get("pathParameters") or {}).get("id")
    if not reg_id:
        return bad_request("id is required")

    try:
        existing = registrations_table().get_item(Key={"registrationId": reg_id}).get("Item")
        if not existing:
            return not_found("Registration not found")
        if existing.get("checkedInAt"):
            return ok({**existing, "alreadyCheckedIn": True})

        now = iso_now()
        registrations_table().update_item(
            Key={"registrationId": reg_id},
            UpdateExpression="SET checkedInAt = :t, checkedInBy = :by, updatedAt = :t",
            ExpressionAttributeValues={":t": now, ":by": caller["actorId"] or "system"},
        )
        audit("participant.checked_in", entity="participant", entity_id=reg_id,
              actor_id=caller["actorId"], actor_label=caller["actorLabel"],
              meta={"name": existing.get("fullName"), "email": existing.get("email"),
                    "phone": existing.get("phone")})
        return ok({**existing, "checkedInAt": now, "checkedInBy": caller["actorId"]})
    except Exception:
        log.exception("checkIn failed")
        return server_error()
